openldap: fix build if openssl or cyrus_sasl are overridden to null, add flag for cyrus_sasl, require openssl

sternenseemann opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Compiling `openldap.override { cyrus_sasl = null; }` results in an evaluation error.


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

cc @lovek323

sternenseemann committed:
openldap: fix build if openssl or cyrus_sasl are overridden to null

Note that compiling without openssl is only a theoretical possibility
since we have no alternative crypto backends supported by the package.

SuperSandro2000 commented on pkgs/development/libraries/openldap/default.nix:
```diff
@@ -54,20 +54,25 @@ stdenv.mkDerivation rec {
     rm -r libraries/*/.libs
     rm -r contrib/slapd-modules/passwd/*/.libs
-  '';
+  '' + stdenv.lib.optionalString (cyrus_sasl != null || openssl != null) (''
+    sed \
+  '' + stdenv.lib.optionalString (cyrus_sasl != null) ''
+        -e 's,-lsasl2,-L${cyrus_sasl.out}/lib -lsasl2,' \
+  '' + stdenv.lib.optionalString (openssl != null) ''
+        -e 's,-lssl,-L${openssl.out}/lib -lssl,' \
+  '' + ''
+        -i $out/lib/libldap.la -i $out/lib/libldap_r.la
+  '');
```
```suggestion
  '' + stdenv.lib.optionalString (cyrus_sasl != null) ''
    sed -e 's,-lsasl2,-L${cyrus_sasl.out}/lib -lsasl2,' -i $out/lib/libldap.la -i $out/lib/libldap_r.la
  '' + stdenv.lib.optionalString (openssl != null) ''
    sed -e 's,-lssl,-L${openssl.out}/lib -lssl,' -i $out/lib/libldap.la -i $out/lib/libldap_r.la
  '');
```
I think the repetition here is worth the readability. Also can we use subsituteInPlace?

sternenseemann replied:
fair enough, pushed a version that should be a bit better in that regard.

sternenseemann commented:
Sorry everyone, rebasing against `master` instead of `staging` by accident completely made github subscribe a lot of people to this PR, you might wanna unsubscribe. 😑

SuperSandro2000 reviewed (commented):
Can we add a flag for this? Overwriting an input with null is not that nice.

sternenseemann committed:
openldap: add flag for cyrus_sasl, require openssl

* instead of cyrus_sasl == null, withCyrusSasl = false is now the proper
  way to disable sasl support for openldap.
* the ability to compile without openssl has been removed, since openssl
  is required as the package needs it for libcrypto.

sternenseemann commented:
> Can we add a flag for this? Overwriting an input with null is not that nice.

Sure.

SuperSandro2000 commented on pkgs/development/libraries/openldap/default.nix:
```diff
@@ -1,4 +1,7 @@
-{ stdenv, fetchurl, openssl, cyrus_sasl, db, groff, libtool }:
+{ stdenv, lib, fetchurl, openssl, db, groff, libtool
+, cyrus_sasl
+, withCyrusSasl ? true
```
```suggestion
, withCyrusSasl ? true, cyrus_sasl
```

Profpatsch replied:
you can’t be serious …

SuperSandro2000 replied:
No, I am just joking and thinking that my review comments are not to be taken serious. sarcasm off.

SuperSandro2000 replied:
You can also move the input below it. Doesn't matter to much.

sternenseemann replied:
have it your way…

SuperSandro2000 commented:
Other than that LGTM.

SuperSandro2000 commented on pkgs/development/libraries/openldap/default.nix:
```diff
@@ -54,19 +56,20 @@ stdenv.mkDerivation rec {
-  # 2. Libraries left in the build location confuse `patchelf --shrink-rpath`
+  # 1. Libraries left in the build location confuse `patchelf --shrink-rpath`
   #    Delete these to let patchelf discover the right path instead.
   #    FIXME: that one can be removed when https://github.com/NixOS/patchelf/pull/98
```
Is that PR already in nixpkgs? Maybe we should address that FIXME some time.

sternenseemann replied:
No that PR is not yet in nixpkgs it seems, tried removing the lines, issue still stands.

sternenseemann replied:
Correction: the feature is in `nixpkgs` but the `patchelf` setup-hook doesn't use the `allowed-rpath-prefixes` option and there's no way to pass it as well.

SuperSandro2000 commented on pkgs/development/libraries/openldap/default.nix:
```diff
@@ -54,19 +56,20 @@ stdenv.mkDerivation rec {
   #    Delete these to let patchelf discover the right path instead.
   #    FIXME: that one can be removed when https://github.com/NixOS/patchelf/pull/98
   #    is in Nixpkgs patchelf.
+  # 2. Fixup broken libtool for openssl and cyrus_sasl respectly if they are not
+  #    overridden to be null.
```
```suggestion
  # 2. If build with withCyrusSasl fixup broken libtool
```

Or something similar but we no longer fixup openssl so the comment is wrong right now.

sternenseemann replied:
Fixed the comment and re added the openssl fix which I removed erroneously yesterday.

SuperSandro2000 commented:
@sternenseemann you did not rebase cleanly and left a merge conflict in.

sternenseemann commented:
Should be fine now.

SuperSandro2000 commented on pkgs/development/libraries/openldap/default.nix:
```diff
@@ -54,19 +56,27 @@ stdenv.mkDerivation rec {
     rm -r libraries/*/.libs
     rm -r contrib/slapd-modules/passwd/*/.libs
+    for f in $out/lib/libldap.la $out/lib/libldap_r.la; do
+<<<<<<< HEAD
+  '' + stdenv.lib.optionalString (cyrus_sasl != null) ''
+      substituteInPlace $f --replace '-lsasl2' '-L${cyrus_sasl.out}/lib -lsasl2'
+  '' + stdenv.lib.optionalString (openssl != null) ''
+=======
+>>>>>>> 52df434b008 (openldap: add flag for cyrus_sasl, require openssl)
```
```suggestion
```

sternenseemann replied:
lol, fixed now.

SuperSandro2000 commented on pkgs/development/libraries/openldap/default.nix:
```diff
@@ -54,19 +56,21 @@ stdenv.mkDerivation rec {
-        -e 's,-lssl,-L${openssl.out}/lib -lssl,' \
-        -i $out/lib/libldap.la -i $out/lib/libldap_r.la
-
     rm -rf $out/var
```
```suggestion
    rm -r $out/var
```
or is this only created with some options? Then we can leave it I guess.

sternenseemann replied:
idk, but doesn't hurt either.

SuperSandro2000 replied:
We usually do not want to silently ignore such things because it could mean upstream changed their layout which we missed.

sternenseemann replied:
Done

SuperSandro2000 reviewed (approved):
Other than that the diff LGTM. @lovek323

sternenseemann committed:
openldap: minor cleanups

* Quote some shell arguments
* Make rm fail if openldap ever ceases to install var to $out

SuperSandro2000 commented:
``openldap.override { withCyrusSasl = true; }``

SuperSandro2000 commented:
Also cross compiles to aarch64-multilib.

SuperSandro2000 commented on pkgs/development/libraries/openldap/default.nix:
```diff
@@ -54,19 +56,21 @@ stdenv.mkDerivation rec {
-        -i $out/lib/libldap.la -i $out/lib/libldap_r.la
-
-    rm -rf $out/var
+    rm -r "$out/var"
```
```suggestion
    rm -r $out/var
```

SuperSandro2000 commented on pkgs/development/libraries/openldap/default.nix:
```diff
@@ -54,19 +56,21 @@ stdenv.mkDerivation rec {
+    rm -r "$out/var"
     rm -r libraries/*/.libs
     rm -r contrib/slapd-modules/passwd/*/.libs
+    for f in "$out/lib/libldap.la" "$out/lib/libldap_r.la"; do
```
```suggestion
    for f in $out/lib/libldap.la $out/lib/libldap_r.la; do
```

sternenseemann replied:
Why _remove_ quoting? This is really bizarre bikeshedding.

SuperSandro2000 replied:
Because it is not required and most parts of nixpkgs is not doing it.

SuperSandro2000 committed:
Apply suggestions from code review

sternenseemann commented:
What GitHub merge method did you use? Apparently this is one where it loses all commit messages except for the (in this case slightly out of date) PR title and number? Would be nice to use one which preserves the commit messages even when squashing.

SuperSandro2000 commented:
Squashing. The is no never method to squash and keep the title. I would need to force push that to your branch.

Profpatsch commented:
@SuperSandro2000  Please don’t squash PRs with sensible commits.

Profpatsch commented:
Rebase them instead

Profpatsch commented:
Alternatively, squash them locally to keep all descriptions and then push the squashed commit before merging. All extended descriptions that @sternenseemann wrote just went away.

SuperSandro2000 commented:
> Alternatively, squash them locally to keep all descriptions and then push the squashed commit before merging. All extended descriptions that @sternenseemann wrote just went away.

git sometimes takes 10 minutes to force push other peoples branches and runs out of memory. I am to lazy to wait for that.
